Feat: Autoconnect and Retry

MrAdder opened the pull request:
Refactored my Auto Connect and Retry feature following the guidance from #12 

Possibility to Fix #7 

I have not tested this, so I apologise if this does not work first try.

Compiled test version for those who can not compile is linked below; however, it does not include #18 

https://ci.appveyor.com/project/MrAdder/vsmr/builds/52123465

For a build that contains #18 as well, use 

https://ci.appveyor.com/project/MrAdder/vsmr/builds/52123426

MrAdder committed:
Feat: Auto Logon | Retry

Refactored and changed the Auto Logon and Retry features

Currently untested

MrAdder committed:
Forgot to add Text

MrAdder committed:
Possible build fix

MrAdder commented:
@19wintersp your insight into this is welcome
as well as @BenWalker01

MrAdder committed:
Updated Appveyor

19wintersp commented:
How does this address #2?

MrAdder commented:
Forgot to commit that part its comming soon™

19wintersp reviewed (changes requested):
Please consider when re-formatting entire files that it can cause merge conflicts and makes reviewing harder. In future, if you do choose to re-format a file, please do it in a separate commit, apply it to all files, _and_ include the formatting rules used (via .clang-format or similar).

There are quite a few instances of implicit casts which should be avoided, but there are worse things already in the codebase. Please do check the compiler output though, as I'm fairly sure at least some of them will be causing warnings to be produced—any such warnings should be addressed before merge.

19wintersp commented:
> Forgot to commit that part its comming soon™

Please do so in a separate PR if so.

I still believe some more consideration of the best course of action wrt #7 is warranted; I'll put some thoughts in the issue thread and would welcome others' input.

MrAdder committed:
Fixes

Unset the isConnecting variable when actually connecting ready for reconnecting attempts

19wintersp commented on vSMR/SMRPlugin.cpp:
```diff
@@ -722,10 +725,13 @@ void CSMRPlugin::OnTimer(int Counter)
 
+	
+	//Auto connection
 	if (AutoConnect && !HoppieConnected && !isConnecting)
```
I don't believe this is properly resolved. The effect now is that, if auto-connect is enabled, repeated connection attempts will be made (without cooldown!) whenever Hoppie's is not connected and a connection attempt is not currently in progress. This also means that, as before, the user is unable to disconnect themselves from Hoppie's if auto-connect is enabled. It also appears the checks which ensure the user is actually connected and controlling have been lost from the last PR.

MrAdder replied:
Thank you for mentioning that I will re-add them in also learning for an exam at the same time, just coding between breaks

19wintersp commented on vSMR/SMRPlugin.cpp:
```diff
@@ -734,7 +740,7 @@ void CSMRPlugin::OnTimer(int Counter)
 		HoppieConnected = false;
-		DisplayUserMessage("CPDLC", "Server", "Logged off!", true, true, false, true, false);
+		DisplayUserMessage("CPDLC", "Server", "Please run .smr connect to login!", true, true, false, true, false);
 	}
```
I ought to have made my previous comments clearer; sorry. This entire block makes little sense, since the effect is to, for any controller with the auto-connect feature disabled when connected to Hoppie's, every second, disconnect them from Hoppie's and inform them to connect. The `else` is also redundant; the conditions are exclusive.

19wintersp reviewed (changes requested):
I implore you to please test your changes before submitting PRs, or at least mark it as a draft if you do not expect reviews yet.

19wintersp commented:
Copying comments from #12:

> I'm also not totally convinced of the utility of continued attempts at connection when doing so has failed already, since this is particularly liable to unnecessarily pestering the server with requests it has already deemed invalid (in the likely scenarios of invalid authentication or an existing conflicting connection), or shouting into the void (if it's a network problem).

As before, I think classifying the cause of failure to connect should be a prerequisite for any attempt at reconnection. What exactly is the intended scenario for which automatic retry on connection failure is to be used?

MrAdder committed:
Update SMRPlugin.cpp

- ReAdded Checks for if the client is a controller or observer
- Added notices Max retries and missing logon code
- fixed the login loop from the autoconnection

MrAdder committed:
Update SMRPlugin.cpp

BenWalker01 commented:
@MrAdder Been busy, sorry. Can't remember where this was up to. Has the auto connect been removed?

MrAdder commented:
Will try to remove that in the morning before work; however, sleep is needed

MrAdder committed:
Disable AutoConnect

MrAdder committed:
Manual Merge of 56b96d3

Manual Merge of 56b96d3

MrAdder committed:
Merge branch 'master' into feat-autoconnect-retry

19wintersp commented:
My previous comment regarding reconnection still stands; I think at minimum an exponential backoff should be implemented for this to be considered at all

BenWalker01 commented:
> Please consider when re-formatting entire files that it can cause merge conflicts and makes reviewing harder. In future, if you do choose to re-format a file, please do it in a separate commit, apply it to all files, _and_ include the formatting rules used (via .clang-format or similar).
> 
> 
> 
> There are quite a few instances of implicit casts which should be avoided, but there are worse things already in the codebase. Please do check the compiler output though, as I'm fairly sure at least some of them will be causing warnings to be produced—any such warnings should be addressed before merge.

Coming back to this, as a few people are working on the file, please keep your changes specific to the part you are working on. Please revert any formatting changes, not relevant to the content you are adding.

For @19wintersp you are correct re the compiler output. I'll fix it eventually (probably).

BenWalker01 commented:
Development has moved to https://github.com/VATSIM-UK/vSMR so I'll close the PR here. Feel free to open a new PR on the fork
